R/sass.R

```diff
@@ -14,8 +14,37 @@
 #'   false-positive cache hit. For this reason, developers are encouraged to
 #'   capture such information in `cache_key_extra` (possibly with
 #'   `packageVersion('myPackage')` if shipping Sass with a package), and users
-#'   may want to disable caching altogether during local development (via
-#'   `sass_cache_set(NULL)`).
+#'   may want to disable caching altogether during local development by calling
+#'   `options(sass.cache=FALSE)`.
+#'
+#'   In some cases when developing and modifying .scss files, [sass()] might not
+#'   detect changes, and keep using cached .css files instead of rebuilding
+#'   them. To be safe, if you are developing a theme with sass, it's best to
+#'   turn off caching by calling `options(sass.cache=FALSE)`.
+#'
+#'   If caching is enabled, [sass()] will attempt to bypass the compilation
+#'   process by reusing output from previous [sass()] calls that used equivalent
+#'   inputs. This mechanism works by computing a _cache key_ from each [sass()]
+#'   call's `input`, `option`, and `cache_key_extra` arguments. If an object
+#'   with that hash already exists within the cache directory, its contents are
+#'   used instead of performing the compilation. If it does not exist, then
+#'   compilation is performed and usual and the results are stored in the cache.
+#'
+#'   If a file that is included using [sass_file()] changes on disk (i.e. its
+#'   last-modified time changes), its previous cache entries will effectively be
+#'   invalidated (not removed from disk, but they'll no longer be matched).
+#'   However, if a file imported using [sass_file()] itself imports other sass
+#'   files using \code{@import}, changes to those files are invisible to the
+#'   cache and you will end up with stale results. Caching should be disabled in
+#'   cases like this.
+#'
+#'   By default, the maximum size of the cache is 40 MB. If it grows past that
+#'   size, the least-recently-used objects will be evicted from the cache to
+#'   keep it under that size. Also by default, the maximum age of objects in the
+#'   cache is one week. Older objects will be evicted from the cache.
+#'
+#'   To clear the default cache, call `sass_default_cache()$reset()`.
```

Comment:
Wouldn't `sass_cache_get()$reset()` be more relevant?

Reply:
`sass_cache_get()` works differently now. It needs to be passed a directory. `sass_default_cache()` is now the function to call to get the default cache.

With the old behavior of `sass_cache_get()`, the first time it was called, it would set a global default directory, and that would last until the R session ends. The problem with that is that it didn't take into account the context where sass is used. For example if we're in a Shiny app, we want to first look in the `cache/` subdir, but if we're not in a Shiny app, we usually want to use the user-level cache dir -- but if we set either of these as the global default, then in another context, we'll be using the wrong cache dir.

So the solution I settled on was to add `sass_context_cache_dir()`, which returns the proper directory depending on the context, and `sass_default_cache()`, which calls that function to find which directory to use.